### PR TITLE
[WIP][SPARK-24497][SQL] Support recursive SQL query

### PR DESCRIPTION
**Pull request by peter-toth**
### What changes were proposed in this pull request?
This PR adds recursive query feature to Spark SQL.

A recursive query is defined using the `WITH RECURSIVE` keywords and referring the name of the common table expression within the query.
The implementation complies with SQL standard and follows similar rules to other relational databases:
- A query is made of an anchor followed by a recursive term.
- The anchor terms doesn't contain self reference and it is used to initialize the query.
- The recursive term contains a self reference and it is used to expand the current set of rows with new ones.
- The anchor and recursive terms must be joined with each other by `UNION` or `UNION ALL` operators.
- New rows can only be derived from the newly added rows of the previous iteration (or from the initial set of rows of anchor terms). This limitation implies that recursive references can't be used with some of the joins, aggregations or subqueries.

Please see `cte-recursive.sql` and `with.sql` for some examples.

Please note that this PR focuses on the minimal working implementation which means:
- SQL recursion is actually loop where the current iteration is computed based on the previous one's result and when an iteration returns no rows the loop is over. The final result is the union of all iteration results. This means that caching intermediate results could speed up the process, but caching was removed from this PR to reduce complexity and can be added back in a follow-up PR.
- A common way to stop SQL recursion is using the LIMIT operator to stop computing more than the required number of rows. LIMIT support was removed from this PR to reduce complexity and can be added back in a follow-up PR.
- Some relational databases are more relaxed in terms how many anchor and recursive terms can be in a recursion. This PR allows the most simple case and allows only 1-1 of them. A follow-up PR can target to relax this limitation.

### Why are the changes needed?
Recursive query is an ANSI SQL feature that is useful to process hierarchical data.

### Does this PR introduce _any_ user-facing change?
Yes, adds recursive query feature.

### How was this patch tested?
Added new UTs and tests in `cte-recursion.sql` and `with.sql`.

### CODE REVIEW

**Comment by peter-toth** [comment]
This is the static version of https://github.com/apache/spark/pull/23531 which means this PR doesn't do adaptive replanning in each recursive iteration, but the advantage is that the implementation comes with simplified code. Common relational DB implementations that doesn't do replanning either.

@maropu this is very close to what you suggested before, if you have some time please review.

@maryannxue I think we can add adaptive support in follow-up PR if needed.

cc @cloud-fan @viirya @gatorsmile

**Comment by maropu** [comment]
The simpler design as a first step looks fine to me. Anyone preferring the adaptive one for this? I think we need to choose which approach to take first. @maryannxue @viirya @gatorsmile @viirya

File: sql/core/src/test/resources/sql-tests/inputs/cte-recursion.sql [hunk 1]
```diff
@@ -0,0 +1,409 @@
+--CONFIG_DIM1 spark.sql.autoBroadcastJoinThreshold=10485760
+--CONFIG_DIM1 spark.sql.autoBroadcastJoinThreshold=-1
+
+--CONFIG_DIM2 spark.sql.adaptive.enabled=false
+--CONFIG_DIM2 spark.sql.adaptive.enabled=true
+
+CREATE TEMPORARY VIEW t AS SELECT * FROM VALUES 0, 1, 2 AS t(id);
+
+-- fails due to recursion isn't allowed with RECURSIVE keyword
```
[the remark this answers is not in the document]
Comment (peter-toth) [reply sql/core/src/test/resources/sql-tests/inputs/cte-recursion.sql]: Thanks @manuzhang. Fixed.

**Comment by peter-toth** [comment]
retest this please

**Comment by peter-toth** [comment]
I'm happy to update this PR if anyone is willing to review. Just let me know...

**Comment by phancey** [comment]
I would like to see recursive queries - why did this not go through? Is there an alternative?

**Comment by sixdimensionalarray** [comment]
Another +1, ran into a use case where recursive queries were needed and alternate technology was utilized to solve because it was not available in Spark in Spark SQL.

**Comment by peter-toth** [comment]
Thanks for the feedback. I will try to rebase this PR on the latest `master` (Spark 3.4) in a few weeks.

**Comment by peter-toth** [comment]
Sorry guys, this is unlikely to land in Spark 3.4, maybe in 3.5...

**Comment by wangyum** [comment]
@peter-toth Could you rebase this PR on the master branch. I have removed the Stale tag.

**Comment by peter-toth** [comment]
> @peter-toth Could you rebase this PR on the master branch. I have removed the Stale tag.

@wangyum, unfortunately this is a very old PR and a lots of changes are needed to make it work on the latest Spark.
I've started rebasing it and fixing it last year, but it requires quite some time and I haven't finished yet.

Let me close the PR for now and reopen once I have a working solution.

**Comment by peter-toth** [comment]
This PR is rebased on latest `master` here: https://github.com/apache/spark/pull/40744

**Pull request closed without merging**